### PR TITLE
Add a subcommand wrapper for `zig cc` and `zig c++`

### LINKED ISSUE #761: Windows support for zig based cross compiling
https://github.com/PyO3/maturin/pull/756#discussion_r776128009

### PR DESCRIPTION
**Pull request by messense**
Closes #761

### CODE REVIEW

File: src/zig.rs [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+            Zig::Cxx { args } => ("c++", args),
+        };
+        // Replace libgcc_s with libunwind
+        let cmd_args: Vec<String> = cmd_args
+            .iter()
+            .map(|arg| {
+                let arg = if arg == "-lgcc_s" {
+                    "-lunwind".to_string()
+                } else if arg.starts_with('@') && arg.ends_with("linker-arguments") {
+                    // rustc passes arguments to linker via an @-file when arguments are too long
```
Comment (messense) [inline src/zig.rs:49]: See https://github.com/rust-lang/rust/issues/41190

File: src/zig.rs [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
```
Comment (konstin) [inline src/zig.rs:49]:
```suggestion
                    // rustc passes arguments to linker via an @-file when arguments are too long
                    // See https://github.com/rust-lang/rust/issues/41190
```

File: src/main.rs [hunk 3]
```diff
@@ -163,6 +163,9 @@ enum Opt {
         #[clap(name = "SHELL", parse(try_from_str))]
         shell: Shell,
     },
+    /// Zig linker wrapper
```
Comment (konstin) [inline src/main.rs:166]: Should be hidden, just like pep517

File: src/zig.rs [hunk 4]
```diff
@@ -0,0 +1,214 @@
+use crate::target::Arch;
```
Comment (konstin) [inline src/zig.rs:1]:
```suggestion
//! Using `zig cc` as c compiler and linker to target a specific glibc/musl version as alternative to the manylinux docker container and for easier cross compiler  

use crate::target::Arch;
```

File: src/main.rs [hunk 5]
```diff
@@ -449,6 +452,9 @@ fn run() -> Result<()> {
                 }
             }
         }
+        Opt::Zig(subcommand) => {
+            subcommand.execute()?;
```
Comment (konstin) [inline src/main.rs:456]:
```suggestion
            subcommand.execute().context("Failed to create zig wrapper script")?;
```

File: src/zig.rs [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+            .spawn()
+            .with_context(|| format!("Failed to run `zig {}`", cmd))?;
+        let status = child.wait().expect("Failed to wait on zig child process");
+        if !status.success() {
+            process::exit(status.code().unwrap_or(1));
+        }
+        Ok(())
+    }
+
+    fn find_zig() -> Result<(String, Vec<String>)> {
```
Comment (konstin) [inline src/zig.rs:74]:
```suggestion
    /// Search for `python -m ziglang` first and for `zig` second. That way we use the zig from `maturin[ziglang]` first, but users or distributions can also insert their own zig
    fn find_zig() -> Result<(String, Vec<String>)> {
```

File: src/zig.rs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        "{} zig {} -- {} $@",
+        current_exe.display(),
+        command,
+        args
+    )?;
+    Ok(())
+}
+
+#[cfg(not(target_family = "unix"))]
+fn write_linker_script(path: &Path, command: &str, args: &str) -> Result<()> {
```
Comment (konstin) [inline src/zig.rs:199]:
```suggestion
/// Write a zig cc wrapper batch script for windows
fn write_linker_script(path: &Path, command: &str, args: &str) -> Result<()> {
```

**Review by konstin — APPROVED** [review]
:+1:

File: src/main.rs (hunk 3 above)
Comment (messense) [reply src/main.rs:166]: It's done here: https://github.com/PyO3/maturin/pull/763/files#diff-49b1f98d9d763bc0e0db9c28ba3dc79ca7413959f017924bc333f7a086d16189R20

Comment (messense) [reply src/main.rs:166]: It turns out clap 3.0 doesn't support set subcommand to hidden in here, it only can be set in subcommand struct/enum definition site, see https://github.com/clap-rs/clap/discussions/3497

**Pull request merged**